api/client/websocketupgradeproto/protocol.go

```diff
@@ -0,0 +1,344 @@
+// protocol, allowing it to gracefully close the connection with a close frame.
+type WebsocketUpgradeConn struct {
+	conn       net.Conn
+	readBuffer []byte
+	readError  error
+	readMutex  sync.Mutex
+	writeMutex sync.Mutex
+
+	logContext            context.Context
+	logger                *slog.Logger
+	once                  sync.Once
+	supportsCloseProtocol bool
+	connType              connectionType
+	protocol              string
+}
+
+// newWebsocketUpgradeConnConfig holds the configuration for creating a new
+// WebsocketUpgradeConn. It includes the context, connection, logger, handshake
+// information, and the type of connection (server or client).
+type newWebsocketUpgradeConnConfig struct {
+	ctx      context.Context
+	conn     net.Conn
+	logger   *slog.Logger
+	hs       ws.Handshake
+	connType connectionType
+}
+
+// newWebsocketUpgradeConn creates a new WebsocketUpgradeConn instance.
+// It initializes the connection with the provided configuration, including
+// the context, connection, logger, and handshake information.
+func newWebsocketUpgradeConn(cfg newWebsocketUpgradeConnConfig) *WebsocketUpgradeConn {
+	return &WebsocketUpgradeConn{
+		logContext:            cfg.ctx,
+		logger:                cfg.logger,
+		conn:                  cfg.conn,
+		supportsCloseProtocol: cfg.hs.Protocol == constants.WebAPIConnUpgradeProtocolWebSocketClose,
+		connType:              cfg.connType,
+		protocol:              cfg.hs.Protocol,
+	}
+}
+
+func (c *WebsocketUpgradeConn) NetConn() net.Conn {
+	return c.conn
+}
+
+func (c *WebsocketUpgradeConn) Read(b []byte) (int, error) {
+	c.readMutex.Lock()
+	defer c.readMutex.Unlock()
+
+	n, err := c.readLocked(b)
+
+	// Timeout errors can be temporary. For example, when this connection is
+	// passed to the kube TLS server, it may get "hijacked" again. During the
+	// hijack, the SetReadDeadline is called with a past timepoint to fail this
+	// Read so that the HTTP server's background read can be stopped. In such
+	// cases, return the original net.Error and clear the cached read error.
+	var netError net.Error
+	if errors.As(err, &netError) && netError.Timeout() {
+		c.readError = nil
+		return n, netError
+	}
+	return n, err
+}
+
+func (c *WebsocketUpgradeConn) readLocked(b []byte) (int, error) {
+	if len(c.readBuffer) > 0 {
+		n := copy(b, c.readBuffer)
+		if n < len(c.readBuffer) {
+			c.readBuffer = c.readBuffer[n:]
+		} else {
+			c.readBuffer = nil
+		}
+		return n, nil
+	}
+
+	// Stop reading if any previous read err.
+	if c.readError != nil {
+		return 0, c.readError
+	}
+
+	for {
+		frame, err := ws.ReadFrame(c.conn)
+		if err != nil {
+			c.readError = err
+			return 0, err
+		}
+
+		// All client frames should be masked.
+		if frame.Header.Masked {
+			frame = ws.UnmaskFrame(frame)
+		}
+
+		switch frame.Header.OpCode {
+		case ws.OpClose:
+			// If we receive a close frame, we should respond with a close frame.
+			if c.supportsCloseProtocol {
```

Comment:
```suggestion
```
should we always reply the close regardless `c.supportsCloseProtocol`?

Reply:
this can happen on both side right? i don't see any downside always replying it at best effort though. This kinda bring anther question, shouldn't `websocketCloseProtocol` also try to send the close regardless `c.supportsCloseProtocol`? even though we know the other side may not be able to reply, we can still try to send it but cut down the deadline.

well, since all upgrades are "migrating" to the new protocol anyway, i guess this doesn't matter eventually.